### DISCUSSION TITLE
Using MQTT-TLS to turn on or off Web Server

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#12333] [Q&A] [answered]
Hello, I have been avoiding MQTT because I am intimidated by it and second I have been busy with trying to perfectly configure my Tasmota Flashed device as “Fort Paranoia”  in my free time.  Now that I think, I have a good secure configuration on my tasmota device I want to play around with MQTT and see if it really makes things more convenient.

Background: I have made few tasmota devices that I am using around my house and even gifting a few to my close family and friends. So far I have been using Philips Hue Emulation feature of tasmota and Homeremote app on my phone to locally control the lights attached to this device. I also have Alexa dot device that automatically recognizes my hue enabled devices and I can use alexa to control the lights from anywhere on the internet. Therefore, I never had any need or desire to use or configure MQTT. To me it seemed rather complicated to put everything together to start using MQTT. Mostly it is the part of setting up an inhouse broker on spare pc/raspberry pi and then maintaining this server is what turned me off. 

But soon I realized that the “Fort Paranoia” device I was configuring had a major problem: In order to have a more secure device it is recommended that web server be turned off (after post compilation customizations have been done and the configuration file saved). In my device, without MQTT I have no way of turning Web Server back on after turning it off! I need a way to turn the web server back on. Without MQTT the only other way is to factory reset the device and then either manually re-enter the custom configuration changes or load the saved configuration file. 

Recently I learned about service called hivemq.cloud which is free online broker for upto 100 client connections using MQTT-TLS. So needless to say, I signed up and was able to connect my freshly flashed tasmota device which I compiled with MQTT-TLS option. But the problem is that I can not figure out how to send commands to my device! I can see what is happening on the device, but I can not for instance turn the lights on or off. 

Here is my Setup: 

1)	Tasmota with MQTT-TLS option turned on and configured to connect to Hivemq.cloud
2)	Hivemq.cloud account
3)	MQTT Explorer App configured to connect to Hivemq.cloud.

I don’t know if it is a firmware configuration problem or a problem with my inability to understand how to issue or format MQTT commands. It could also be that MQTT explorer is meant for viewing and monitoring only and does not allow to send commands. But there is a pane on right side that says publish, so I am assuming it is possible to send commands to my device and I tried sending the command “Tasmota_AEE863/Power = 1” but nothing happened. 
(1)	Can someone please help me with formatting a simple turn light on and turn light off command that I can issue through this (Mqtt Explorer) app? 
(2)	Or suggest another MQTT app that is easier to use? 
(3)	 After that I will move on to figuring out how to format and send a command to turn on the “web Server”. Or we can do questions (1) above with “web server on – off” command itself 😊?

Here are some pictures of above setup:

![Screenshot (153)](https://user-images.githubusercontent.com/69470381/121476093-e152ef00-c9e3-11eb-82d7-19e9e65468d0.png)
![Screenshot (154)](https://user-images.githubusercontent.com/69470381/121476137-eca61a80-c9e3-11eb-9002-53909872ad7d.png)
![Screenshot (156)](https://user-images.githubusercontent.com/69470381/121476188-fc256380-c9e3-11eb-8545-e2d3f616393e.png)
![Screenshot (155)](https://user-images.githubusercontent.com/69470381/121476262-10696080-c9e4-11eb-8cb1-37c1cba9e53e.png)
![Screenshot (157)](https://user-images.githubusercontent.com/69470381/121476294-195a3200-c9e4-11eb-85ec-68a7dde342d7.png)

Regards,

Navi

**Comment by s-hadinger** [comment, marked as answer]
You need to separate the command and the payload. Please read here: https://tasmota.github.io/docs/MQTT/

For ex: Send to topic `cmnd/Tasmota_AEE863/Power` with payload `1`.

Are the device connecting correctly to the broker via TLS?

**Reply by NsinghP** [reply]
> For ex: Send to topic `cmnd/Tasmota_AEE863/Power` with payload `1`.

Ok so I published 

cmnd/tasmota_AEE863/wificonfig 4 

after selecting JSON on the MQTT explorer client. But I don't see anything happeing! Shouldnt the tasmota device web serial console show activity when i Publsih this command from the MQTT explorer client? Could it be that i need to compile my binary again because I had custom define string of #use tasmota discovery?

Edit: when i enter a command on the tasmota device web console it shows up in the MQTT client. But not vice versa.

**Reply by sfromis** [reply]
Serial console logging will by default be disabled 10 minutes after boot. And you may have a different logging level for serial vs web.

**Reply by NsinghP** [reply]
> 
> 
> Serial console logging will by default be disabled 10 minutes after boot. And you may have a different logging level for serial vs web.

I have turned off serial console in binary. But I am trying to see if my mqtt explorer issued commands are having any effect by checking the web console. For example  I just published the command cmnd/tasmota_AEE863/power off. but nothing happened. 

EDIT: when I entered the same command on web console the relay1 turned off:

13:49:12.090 CMD: cmnd/tasmota_AEE863/power off
13:49:12.467 MQT: stat/tasmota_AEE863/RESULT = {"POWER1":"OFF"}
13:49:12.804 MQT: stat/tasmota_AEE863/POWER1 = OFF

and it was also reflected in MQTT explorer:

![Screenshot (158)](https://user-images.githubusercontent.com/69470381/121491345-7c9f9080-c9f3-11eb-9ac1-4e76212f953b.png)

**Reply by sfromis** [reply]
`cmnd/tasmota_AEE863/power off` is technically an invalid command in the console, it should just be `power off` to not give a misleading impression of what happens. However, the nonsense parts are dropped, hiding the error. Exactly how did you fill in topic and in the separate text area payload in MQTT Explorer?

Using `Weblog 3` and restarting, you can in the Tasmota console see exactly what topics it subscribes to, to check if it matches what you use.

**Reply by NsinghP** [reply]
> Using `Weblog 3` and restarting, you can in the Tasmota console see exactly what topics it subscribes to, to check if it matches what you use.

This is what i get using weblog 3:

15:04:27.291 MQT: Connected
15:04:27.659 MQT: tele/tasmota_AEE863/LWT = Online (retained)
15:04:27.823 MQT: cmnd/tasmota_AEE863/POWER = 
15:04:27.825 MQT: Subscribe to cmnd/tasmota_AEE863/#
15:04:27.990 MQT: Subscribe to cmnd/tasmotas/#
15:04:28.154 MQT: Subscribe to cmnd/DVES_AEE863_fb/#

How do I use this information to format a command that I can publish from MQTT explorer?

**Reply by sfromis** [reply]
Looks standard. You did not answer my question about how *exactly* you filled in the two separate fields in MQTT Explorer.

**Reply by NsinghP** [reply]
> 
> 
> Looks standard. You did not answer my question about how _exactly_ you filled in the two separate fields in MQTT Explorer.

All I did is hit the publish button after entering the line with blue background

![MQTT exp](https://user-images.githubusercontent.com/69470381/121506884-c7c0a000-ca01-11eb-9ede-565902b3b806.jpeg)

**Reply by sfromis** [reply]
That is so wrong on all accounts. Tasmota does not subscribe to `tele/tasmota_AEE863/STATE` which you use as topic. Simple example is topic `cmnd/tasmota_AEE863/power` and payload `on`. The command name goes in the topic, and argument in the payload.

If you want multiple commands, topic `cmnd/tasmota_AEE863/Backlog` and a semicolon-delimited list of commands in payload.

To repeat, a Tasmota command **never** includes the `cmnd/tasmota_AEE863/` (or similar) part, command name is one word without blanks or slashes.

**Reply by NsinghP** [reply]
Ok when I  give the command "cmnd/tasmota_AEE863/power on" by hitting the publish button: 

![Screenshot (159)](https://user-images.githubusercontent.com/69470381/121513370-694af000-ca08-11eb-8d6e-edd035f81fe6.png)


Nothing happens to the relay state, But Web console shows following:

16:23:07.788 SRC: MQTT
16:23:07.791 CMD: Grp 0, Cmnd 'POWER ON', Idx 1, Len 0, Data ''
16:23:08.287 MQT: stat/tasmota_AEE863/RESULT = {"Command":"Unknown"}

**Reply by sfromis** [reply]
Stop what you're doing, and start actually **reading** what I wrote. You ignored most of it

**Reply by NsinghP** [reply]
> Stop what you're doing, and start actually **reading** what I wrote. You ignored most of it

Sorry, attached wrong screen shot then edited with the latest screenshot :(

**Reply by sfromis** [reply]
My comment related to your latest screenshot. You are just ignoring the repeated reminders about splitting into topic and argument, in the separate fields.

**Reply by sfromis** [reply]
If you do not understand some words, ask instead of ignoring them.

**Reply by NsinghP** [reply]
> 
> 
> If you do not understand some words, ask instead of ignoring them.

LOL, I swear to god, the first time I tried separating the two, it did'nt work. It is magic, now it works:

16:30:46.085 SRC: MQTT
16:30:46.088 CMD: Grp 0, Cmnd 'POWER', Idx 1, Len 3, Data 'off'
16:30:46.492 MQT: stat/tasmota_AEE863/RESULT = {"POWER1":"OFF"}
16:30:47.060 MQT: stat/tasmota_AEE863/POWER1 = OFF
 
Edit: and it was so satisfying to see the relay light go off!  Hey thanks for your patience. Now I will mess around with some other commands before turning off the web server and turning it back on. 

Edit2: Now I am able to turn off the webserver: "webserver 0" and turn it back on in user/admin mode using MQTT "webserver 1" and "webserver 2" commands. Special thanks to sfromis for his time and effort.
